### PR TITLE
Add the Design pattern article

### PR DESCRIPTION
**Pull request by elmansour1**

### CODE REVIEW

File: architecture/design-pattern/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+tags:
+  - Design
+  - pattern
+  - software architecture
+date: 18-02-2024
+slug:
+---
+
+# Introduction
+![fearuredImage](./Images/design-pattern.png)
```
Comment (dassiorleando) [inline architecture/design-pattern/README.md:13]: You have a **"duplicated featured image"** in your files, and the quality of the one you used could be improved.

For the featured images, we use visually appealing ones like the previous images published on the blog.
https://numericaideas.com/blog/

I can provide it to you if you can't do something similar.

**Review by dassiorleando — CHANGES_REQUESTED** [review]
@elmansour1 this is a great content, all the terms are nicely explained, but we can slightly improve the quality of the article before releasing it.

Thanks in advance to consider the above-mentioned points.

Looking forward to your updates.

Cheers!

File: architecture/design-pattern/README.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+## Historical:
+Design pattern is a concept that goes back a bit. It originated with the publication of the book A Pattern Language: Towns, Buildings, Construction (Center for Environmental Structure Series) by Christopher Alexander. It is a book that presents solutions that can be used to design an urban environment. The term was picked up in 1994 and was adapted by a group nicknamed the Gang of Four (Erich Gamma, Richard Helm, Ralph Johnson, and John Vlissides) into the software field where they published a book Design Patterns: Catalog of Design Patterns. reusable design. Hence the origin of the word Design Pattern in the field of IT. In this book, they presented 23 solution models to use when faced with a number of recurring problems in the world of computer programming. Since then, the use of design patterns has become an essential part of software programming.
+
+## Classification of design patterns
```
[the remark this answers is not in the document]
Comment (elmansour1) [reply architecture/design-pattern/README.md:24]: ok

File: architecture/design-pattern/README.md (hunk 1 above)
Comment (elmansour1) [reply architecture/design-pattern/README.md:13]: Yes, can you provide it to me please ?

File: architecture/design-pattern/README.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+There are several categories of design patterns, including creational, structural, and behavioral patterns. In this section, we will briefly present each category:
+
+### Creational pattern:
+These models focus on the principles of object creation, allowing the creation of flexible and isolated objects. This is to avoid strong coupling between objects. There are 5 in the literature. Below we will cite some of them:
+
+#### Singleton:
```
[the remark this answers is not in the document]
Comment (dassiorleando) [reply architecture/design-pattern/README.md]: > i remove the ":" at the end of each title ?

Yes let's remove them and make sure each pattern has a full tile like **"Singleton Design Pattern"** ... not only ~~"Singleton"~~.

[the remark this answers is not in the document]
Comment (elmansour1) [reply architecture/design-pattern/README.md]: I think it's already ok here

File: architecture/design-pattern/README.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
+---
```
Comment (dassiorleando) [inline architecture/design-pattern/README.md:1]: Could we remove this section and use the following:

# Design Patterns&nbsp;[Hits](https://numericaideas.com/blog/design-patterns)

**This article was originally written by "Faouziel Mansour" on the blog**: https://numericaideas.com/blog/design-patterns


It's almost the same for other articles, here's an example: https://github.com/numerica-ideas/community/edit/master/java/maven-for-beginners/README.md

File: architecture/design-pattern/README.md (hunk 1 above)
Comment (dassiorleando) [reply architecture/design-pattern/README.md:13]: @elmansour1 I still see the duplicated featured images in the images folder.

TODO:
- The images folder should be named **images**, not Images. small caps.
- The article folder should be named **design-patterns**, not design-pattern. plural form.

As attached file, the featured image to use:

Design Patterns

File: architecture/design-pattern/README.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+This pattern, as its name suggests, is intended to be strategic. It allows defining a family of dynamically interchangeable algorithms and encapsulates each of them, allowing them to be used interchangeably depending on the context. This pattern addresses the problems of using a single object with multiple treatments without violating the SOLID principles.
+
+#### Command Design Pattern
+It allows you to encapsulate a request as an object, that is to say, it allows you to separate the action code and the action itself. This pattern allows you to design an application without knowing the characteristics of an object with which your application is called upon to communicate. Much more technically, this pattern allows clients to configure clients with different requests, queue or log requests, and support cancelable operations.
+
+
+These are just a few examples of design templates, and there are many more templates available. Each model addresses a specific design problem and provides a proven solution. It is important to choose the appropriate model based on the problem at hand and the specific requirements of your software project.
+
+## Why Design Patterns?
```
Comment (dassiorleando) [inline architecture/design-pattern/README.md:67]: I'm not sure if it's good to put the **"Why Design Patterns"** part after we have listed them on the top.
What do you think about moving it after the **"Historical"** section?

File: architecture/design-pattern/README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+#### Over-engineering
+Design patterns should be used judiciously and only when they are truly necessary. Applying design patterns without a clear need can lead to over-engineering, where the code becomes too complex and difficult to maintain. It's important to strike a balance between using design patterns to solve specific problems and keeping the code base simple and straightforward.
+
+#### Learning curve
+Design patterns require developers to have a solid understanding of their concepts and implementation details. Learning and mastering design patterns can take time and effort, especially for developers who are new to software design principles. This learning curve can slow down the development process at first.
+
+## Conclusion
+Overall, design patterns contribute to the development of high-quality, maintainable, flexible, and scalable software systems. They provide proven solutions to common design problems and promote best practices in software development. However, it is important to use design patterns judiciously and consider the specific requirements and constraints of each project to ensure that the benefits of design patterns outweigh the potential drawbacks.
+
```
Comment (dassiorleando) [inline architecture/design-pattern/README.md:101]: This text should go before the conclusion:

———————

We have just started our journey to build a network of professionals to grow even more our free knowledge-sharing community that’ll give you a chance to learn interesting things about topics like cloud computing, software development, and software architectures while keeping the door open to more opportunities.

Does this speak to you? If **YES**, feel free to [Join our Discord Server](https://discord.numericaideas.com) to stay in touch with the community and be part of independently organized events.

———————

This one after the conclusion:

Thanks for reading this article. Like, recommend, and share if you enjoyed it. Follow us on [Facebook](https://www.facebook.com/numericaideas), [Twitter](https://twitter.com/numericaideas), and [LinkedIn](https://www.linkedin.com/company/numericaideas) for more content.


Here's an existing example: https://github.com/numerica-ideas/community/edit/master/java/maven-for-beginners/README.md

**Review by dassiorleando — CHANGES_REQUESTED** [review]

File: architecture/design-pattern/README.md (hunk 4 above)
Comment (elmansour1) [reply architecture/design-pattern/README.md:1]: ok

File: architecture/design-pattern/README.md (hunk 5 above)
Comment (elmansour1) [reply architecture/design-pattern/README.md:67]: Ah yeah. it's ok. i move after historical section

File: architecture/design-pattern/README.md (hunk 6 above)
Comment (elmansour1) [reply architecture/design-pattern/README.md]: ok

**Review by dassiorleando — APPROVED** [review]
@elmansour1 great article, I'm merging it now.
Thanks for the hard work delivering this.

**Pull request merged**